tools/kcli/kcli/utils.py

```diff
@@ -0,0 +1,230 @@
+import os
+import logging
+import re
+import sys
+
+import configure
+import yaml
+
+import kcli.yamls
+from kcli import exceptions
+from kcli import logger
+
+LOG = logger.LOG
+
+
+# TODO: check if can be moved into Lookup
+def dict_lookup(dic, key, *keys):
+    """lookup and return value of a nested key from a given dictionary
+
+    to get the value of a nested key, all ancestor keys should be given as
+    method's arguments
+
+    example:
+      dict_lookup({'key1': {'key2': 'val'}}, 'key1.key2'.split('.'))
+
+    :param dic: dictionary object to get the key's value from
+    :param key: key / first key in a keys' chain
+    :param keys: sub keys in a keys' chain
+    :return: value of a gives keys
+    """
+    if LOG.getEffectiveLevel() <= logging.DEBUG:
+        calling_method_name = sys._getframe().f_back.f_code.co_name
+        current_method_name = sys._getframe().f_code.co_name
+        if current_method_name != calling_method_name:
+            full_key = list(keys)
+            full_key.insert(0, key)
+            LOG.debug("looking up the value of \"%s\"" % ".".join(full_key))
+
+    if key not in dic:
+        if isinstance(key, str) and key.isdigit():
+            key = int(key)
+        elif isinstance(key, int):
+            key = str(key)
+
+    if keys:
+        return dict_lookup(dic.get(key, {}), *keys)
+
+    try:
+        value = dic[key]
+    except KeyError:
+        raise exceptions.IRKeyNotFoundException(key, dic)
+
+    LOG.debug("value has been found: \"%s\"" % value)
+    return value
+
+
+def dict_insert(dic, val, key, *keys):
+    """insert a value of a nested key into a dictionary
+
+    to insert value for a nested key, all ancestor keys should be given as
+    method's arguments
+
+    example:
+      dict_lookup({}, 'val', 'key1.key2'.split('.'))
+
+    :param dic: a dictionary object to insert the nested key value into
+    :param val: a value to insert to the given dictionary
+    :param key: first key in a chain of key that will store the value
+    :param keys: sub keys in the keys chain
+    """
+    if not keys:
+        dic[key] = val
+        return
+
+    if key not in dic:
+        dic[key] = {}
+
+    dict_insert(dic[key], val, *keys)
+
+
+# TODO: remove "settings" references in project
+def validate_settings_dir(settings_dir=None):
+    """Checks & returns the full path to the settings dir.
+
+    Path is set in the following priority:
+    1. Method argument
+    2. System environment variable
+    3. Settings dir in the current working dir
```

Comment:
I don't think this should be an option. we shouldn't rely on current working dir for this.